torch/distributed/pipeline/sync/pipe.py

```diff
@@ -133,6 +133,48 @@ def forward(self, *inputs):
         return inputs
 
 
+class WithDevice(nn.Module):
+    """
+    Wraps an nn.Module which is part of nn.Sequential passed into :class:`Pipe`
+    that overrides the device for that module. In cases where :class:`Pipe`
+    can't implicitly determine the device for the module and places it on CPU,
+    this wrapper can be used to override the implicit behavior and explicitly
+    specify which device a module should run on.
+
+    The provided module is also moved to the given device via ``.to(device)``
+    by :class:`Pipe`
+
+    Args:
+        module(:class:`torch.nn.Module`): The module to be wrapped.
+        device(:class:`torch.device`): The device to run the module on.
+
+    Example::
+        >>> fc1 = nn.Linear(16, 8).cuda(0)
+        >>> fc2 = nn.Linear(8, 4).cuda(1)
+        >>> dropout = nn.Dropout()
+        >>>
+        >>> # Dropout does not have any parameters/buffers, but we want to
```

Comment:
I wonder what's the expected behavior here. If `dropout` does not have a device, should we just (recursively) inherit the device from the previous layer? So that users don't need to manually specify this?

Reply:
+1

The goal here is to avoid the extra data transfer, so I also think no need to manually specify a device.

Another future use case where`WithDevice` may not work well is that, if the precursor layer is a tensor with lazy placement (`device="meta"`), then the current pure-computation layer can only be placed on a meta device, but I am afraid that the heuristic of co-locating two layers still won't be easily expressed by `WithDevice`.

Reply:
I thought about this option as well, although this wouldn't work for the first partition in the sequence where we don't have any previous information to rely on. Also, I do feel it's probably better have an explicit API to deal with this rather than having many implicit rules.

For example, what if the user actually wanted a GPU to CPU transfer for a particular layer. Or if there was a sequence of layers with no params/buffers and they want them to run on different devices to spread out computation evenly?

Reply:
> Another future use case whereWithDevice may not work well is that, if the precursor layer is a tensor with lazy placement (device="meta"), then the current pure-computation layer can only be placed on a meta device, but I am afraid that the heuristic of co-locating two layers still won't be easily expressed by WithDevice.

When the Pipe eventually initializes everything, we would have concrete devices for each partition and not "meta" at that point.

Reply:
> e, what if the user actually wanted a GPU to CPU transfer for a particular layer. Or if there was a sequence of layers with no params/buffers and they want them to run on different devices to spread out computation evenly?



> I thought about this option as well, although this wouldn't work for the first partition in the sequence where we don't have any previous information to rely on. Also, I do feel it's probably better have an explicit API to deal with this rather than having many implicit rules.

Is it possible that the first layer is a pure computation layer? As long as the first layer in the first partition is not pure computation layer, shouldn't we always be able to find a valid device from the previous layers?

> For example, what if the user actually wanted a GPU to CPU transfer for a particular layer. Or if there was a sequence of layers with no params/buffers and they want them to run on different devices to spread out computation evenly?

Then that computation must be much more expensive than H2D ops. I agree `WithDevice` be more valuable in this case, but isn't this a uncommon use case? Maybe we can let the placement be the same as the previous layer by default, and provide `WithDevice` for extra flexibility?

Reply:
> Is it possible that the first layer is a pure computation layer? As long as the first layer in the first partition is not pure computation layer, shouldn't we always be able to find a valid device from the previous layers?

It could be possible, for example we didn't anticipate that users would want a separate compute layer without any params/buffers as a different stage. Also, I think there is also a more general case where we could have a sequence of compute layers which need to be on different devices, in that case we do need `WithDevice`.

> Maybe we can let the placement be the same as the previous layer by default, and provide WithDevice for extra flexibility?

If we have to provide `WithDevice` anyways, I feel we should just have this instead of having a variety of implicit rules around how devices are handled.